nimare/meta/utils.py

```diff
@@ -119,33 +123,69 @@ def compute_kda_ma(
     )
     kernel = cube[:, np.sum(np.dot(np.diag(vox_dims), cube) ** 2, 0) ** 0.5 <= r]
 
-    all_coords = []
-    # Loop over experiments
-    for i_exp, _ in enumerate(exp_idx_uniq):
-        # Index peaks by experiment
-        curr_exp_idx = exp_idx == i_exp
-        peaks = ijks[curr_exp_idx]
+    if sum_across_studies:
+        all_values = np.zeros(shape, dtype=np.int32)
+
+        # Loop over experiments
+        for i_exp, _ in enumerate(exp_idx_uniq):
+            # Index peaks by experiment
+            curr_exp_idx = exp_idx == i_exp
+            peaks = ijks[curr_exp_idx]
+
+            sphere_coords = _convolve_sphere(kernel, peaks, np.array(shape))
+
+            # Go from sparse to dense
+            study_values = np.zeros(shape, dtype=np.int32)
+
+            if sum_overlap:
+                sphere_coords = unique_rows(sphere_coords)
+            else:
+                study_values[sphere_coords[:, 0], sphere_coords[:, 1], sphere_coords[:, 2]] = value
+
+            # Sum across studies
+            all_values += study_values
+
+        # Set voxel outside the mask to zero.
+        all_values[~mask_data] = 0
+
+        # Go from dense to sparse
+        nonzero_idx = np.where(all_values > 0)
+        all_coords = np.vstack(nonzero_idx)
+        value = all_values[nonzero_idx]
+
+        # Add dummy indicator to return 4D sparse array
+        exp_indicator = np.zeros(all_coords.shape[1])
+        all_coords = np.insert(all_coords, 0, exp_indicator, axis=0)
+
+        kernel_data = sparse.COO(all_coords, data=value, has_duplicates=False, shape=kernel_shape)
+
+    else:
+        all_coords = []
+        # Loop over experiments
+        for i_exp, _ in enumerate(exp_idx_uniq):
+            # Index peaks by experiment
+            curr_exp_idx = exp_idx == i_exp
+            peaks = ijks[curr_exp_idx]
 
-        # Convolve with sphere
-        all_spheres = _convolve_sphere(kernel, peaks, np.array(shape))
+            # Convolve with sphere
+            all_spheres = _convolve_sphere(kernel, peaks, np.array(shape))
 
-        if not sum_overlap:
-            all_spheres = unique_rows(all_spheres)
+            if not sum_overlap:
+                all_spheres = unique_rows(all_spheres)
 
```

Comment:
still need this bit of code here for the original process, no?

```suggestion
        sphere_idx_inside_mask = np.where(mask_data[tuple(all_spheres.T)])[0]
        all_spheres = all_spheres[sphere_idx_inside_mask, :]
```